synapse/events/snapshot.py

```diff
@@ -209,11 +196,10 @@ def deserialize(storage: "StorageControllers", input: JsonDict) -> "EventContext
             storage=storage,
             state_group=input["state_group"],
             state_group_before_event=input["state_group_before_event"],
-            prev_group=input["prev_group"],
+            state_group_deltas=_decode_state_group_delta(input["state_group_deltas"]),
```

Comment:
I'm not really sure when we serialize / deserialize these, do we need to do anything special during an update / rollout to account for if we serialize with the old `prev_group` and `delta_ids` fields, but then try to deserialize with the new code? Is that possible?

Reply:
If my reading of the code is correct, we serialize/de-serialize when sending events/contexts over replication, so I don't?? think there would be a situation where one replication endpoint would receive an incompatible context - contexts are ephemeral and not stored in the DB so I don't think upgrading would cause an issue here?

The only place I could see this being an issue is if we send event contexts out over federation and a Synapse running this version sends some event contexts to a Synapse running an older version (or vice-versa) but AFAICT we don't send event contexts out over federation, although I should probably ask to double-check. Please let me know if my logic isn't sound here, I am thinking through it real-time.

Reply:
I seem to only see things like [this](https://github.com/matrix-org/synapse/blob/develop/docs/upgrade.md#changes-to-the-account-data-replication-streams) that say replication will pause until things are upgraded. Not sure if we've ever done a 'you must restart the world' one before (I thought we did, but I can't find it...)

Reply:
I think using a similar warning like what you linked would work, as the issue is the same - the upgrade is backwards/forwards incompatible and the workers/master must all be on the same compatible version (and actually I think? this also implies a restart as don't you need to restart to upgrade?)

Reply:
I guess it would be a huge pain to have a backwards compat path since we would also need a way to *roll back* when doing this, which would mean landing the ability to decode the newer map first, waiting a version or two and then landing the new code.

tl;dr Can we document what needs to be done in the upgrade notes?

Reply:
I've added some words on this, let me know if it's clear enough - in particular I wonder if I was too specific about the changes.